pkg/csi_driver/controller.go

```diff
@@ -101,15 +106,66 @@ func (s *controllerServer) CreateVolume(ctx context.Context, req *csi.CreateVolu
 			return nil, status.Error(codes.AlreadyExists, err.Error())
 		}
 	} else {
+
+		// Get the reserved-IPV4-cidr from the request. If value is not provided, reservedIPV4Range is defaulted to ""
+		reservedIPV4CIDR, ok := req.GetParameters()[paramReservedIPV4CIDR]
+		if ok {
+
+			// Prevents lock acquisition in case the CIDR is invalid
+			validCIDR := s.config.ipAllocator.ValidateCIDR(reservedIPV4CIDR)
+			if !validCIDR {
+				return nil, err
+			}
+
+			reservedIPRange, err := s.reserveIPV4Range(ctx, newFiler, reservedIPV4CIDR)
+			if err != nil {
+				return nil, err
+			}
+
+			// Adding the reserved IP range to the instance object
+			newFiler.Network.ReservedIpRange = reservedIPRange
+		}
+
 		// Create the instance
 		filer, err = s.config.fileService.CreateInstance(ctx, newFiler)
 		if err != nil {
 			return nil, status.Error(codes.Internal, err.Error())
```

Comment:
We should unhold the IP in both success and error cases.  A defer can be used after the Hold()

Reply:
In case of error, instance has not been created yet. So other threads making the list  call would not get the the IP as reserved and can end up holding the reserved IP as it is no longer blocked.  Returning a response from CreateVolume ensures that the instance has been created and would be available in subsequent list calls. Let me know if I am missing something in my analysis here.

Reply:
If you reserve the IP, and then CreateInstance fails, right now, it seems like the behavior is to keep it reserved so no other requests can use it.  We want the opposite behavior.  This may also be a good scenario to unit test.

Reply:
In this case we would need to identify whether the CreateInstance operation is running in the background and the error is due to some other source(ex timeout) or the operation itself aborted. In the previous case, we should not release the IP but in the later case we should,

Reply:
I think both cases you can release the IP.  IIUC,

1. If operation is running in the background, then when you `ListInstances`, you will get the IP
2. If operation aborted, the IP didn't get consumed by the Filestore service, so it is safe to release